### PR TITLE
[video][ios] allow disabling live text interaction for videos

### PR DESCRIPTION
**Pull request by fobos531**
# Why

https://discord.com/channels/695411232856997968/1234939321316802672/1256120689480568883

We have it for `expo-image`, so it makes perfect sense to have it here well. The only interesting difference is that this is enabled by default by Apple, whereas with images it is turned off. I think that in this case it makes sense to follow Apple's defaults.

# How

https://developer.apple.com/documentation/avkit/avplayerviewcontroller/allowsvideoframeanalysis

# Test Plan

Added test case to bare-expo

# Checklist

- [X] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [X] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

Demo video:

**NOTE**:  As you'll see in the demo, first I kept it enabled and then second time I left it disabled. However, you will notice that before the second test case, I exited the screen and re-entered it. This is because I noticed that if I disable the analyzer (`allowsVideoFrameAnalysis`) AFTER some text was already scanned causes some issues. Firstly, the "show text" in the native controls is still visible. Also, if I try to tap it, this messes up the native controls to the point they don't show up at all until I exit the screen and enter it again. It's almost like the video view would need to get completely "refreshed". Any pointers on how to fix this would be greatly appreciated cc @behenate   

https://github.com/expo/expo/assets/19533289/a54e7203-83eb-41ee-8d1d-132600578017

### CODE REVIEW

**Comment by behenate** [comment]
Considering that all of the other controls react correctly to changing values, I'm wondering why this one behaves differently. Maybe it's a bug on iOS's side 🧐. I haven't found a way to make it refresh, but I'll keep looking.

File: packages/expo-video/src/VideoView.types.ts [hunk 1]
```diff
@@ -86,4 +86,12 @@ export interface VideoViewProps extends ViewProps {
    * @platform ios 14.2+
    */
   startsPictureInPictureAutomatically?: boolean;
+
+  /**
+   * Specifies whether to perform video frame analysis (Live Text in videos). Check official [Apple documentation](https://developer.apple.com/documentation/avkit/avplayerviewcontroller/allowsvideoframeanalysis) for more details.
+   * @default true
+   * @platform ios 16.0+
+   * @platform macos 15.0+
```
Comment (behenate) [inline packages/expo-video/src/VideoView.types.ts:94]:
```suggestion
```
We can remove this line, and keep only the iOS one.

File: packages/expo-video/ios/VideoView.swift [hunk 2]
```diff
@@ -29,6 +29,18 @@ public final class VideoView: ExpoView, AVPlayerViewControllerDelegate {
...
 
+  #if os(tvOS)
+  let allowsVideoFrameAnalsysis = false
+  #else
+  var allowsVideoFrameAnalsysis: Bool = true {
+    didSet {
+      if #available(iOS 16.0, macCatalyst 18.0, *) {
+        playerViewController.allowsVideoFrameAnalysis = allowsVideoFrameAnalsysis
+      }
+    }
```
Comment (behenate) [inline packages/expo-video/ios/VideoView.swift:40]: We found an ugly workaround and couldn't come up with anything better. At least it seems to work well.

```suggestion
  var allowsVideoFrameAnalsysis: Bool = true {
    didSet {
      if #available(iOS 16.0, macCatalyst 18.0, *) {
        forceAllowsVideoFrameAnalysis(allowsVideoFrameAnalsysis)
      }
    }
```
And add a function under `func stopPictureInPicture()..` 

```swift
  @available(iOS 16.0, *)
  private func forceAllowsVideoFrameAnalysis(_ allowsVideoFrameAnalsysis: Bool) {
    playerViewController.allowsVideoFrameAnalysis = allowsVideoFrameAnalsysis

    // Setting the `allowsVideoFrameAnalysis` to false doesn't update the UI. We can force the desired behaviour by quickly toggling the property.
    // Setting it to true clears existing requests, which updates the UI, hiding the button, then setting it to false before it detects any text keeps it
    // in the desired state.
    // Tested in iOS 17.5
    if (!allowsVideoFrameAnalsysis) {
      playerViewController.allowsVideoFrameAnalysis = true
      playerViewController.allowsVideoFrameAnalysis = false
    }
  }
  ```

File: apps/native-component-list/src/screens/Video/VideoScreen.tsx [hunk 3]
```diff
@@ -241,6 +250,18 @@ export default function VideoScreen() {
...
+            <TitledSwitch
+              title="Allow video frame analysis (Live Text interaction for video)"
+              value={allowsVideoFrameAnalysis}
+              setValue={setAllowsVideoFrameAnalysis}
+              style={styles.switch}
+              titleStyle={styles.switchTitle}
+            />
+          </View>
+        )}
+
```
Comment (behenate) [inline apps/native-component-list/src/screens/Video/VideoScreen.tsx:264]:
```suggestion
```
Let's group it with the last button row

```js
  <View style={styles.row}>
    <TitledSwitch
      title="Show now playing notification"
      value={showNowPlayingNotification}
      setValue={updateShowNowPlayingNotification}
      style={styles.switch}
      titleStyle={styles.switchTitle}
    />
    {Platform.OS === 'ios' && (
      <TitledSwitch
        title="Allow video frame analysis (Live Text interaction for video)"
        value={allowsVideoFrameAnalysis}
        setValue={setAllowsVideoFrameAnalysis}
        style={styles.switch}
        titleStyle={styles.switchTitle}
      />
    )}
  </View>

**Review by behenate — CHANGES_REQUESTED** [review]
Thank you for the contribution! Left a few suggestions as well as a workaround for the UI issue.

File: packages/expo-video/ios/VideoView.swift (hunk 2 above)
Comment (fobos531) [reply packages/expo-video/ios/VideoView.swift]: That works great! Nice find @behenate

File: packages/expo-video/ios/VideoModule.swift [hunk 4]
```diff
@@ -72,6 +72,12 @@ public final class VideoModule: Module {
         #endif
       }
 
+      Prop("allowsVideoFrameAnalysis") { (view, allowsVideoFrameAnalysis: Bool?) in
```
[the remark this answers is not in the document]
Comment (fobos531) [reply packages/expo-video/ios/VideoModule.swift:75]: Makes sense, done

**Comment by fobos531** [comment]
The changes have been implemented - I have also implemented Tomasz's feedback by removing the logic from the view and adding it directly in the module. Let me know if any further changes are required.

**Review by tsapeta — APPROVED** [review]
Great, looks good to me now! Thanks for the contribution! 🙏

**Pull request merged**